File: static/csharp/Validation.cs
// This file is a part of the Polygen common utility library.
// It provides the validation system for checking data integrity after loading.
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace Polygen.Common
{
    /// <summary>
    /// Severity level for validation errors.
    /// </summary>
    public enum ValidationSeverity
    {
        /// <summary>
        /// Critical error that should prevent data usage.
        /// </summary>
        Error,

        /// <summary>
        /// Warning that may indicate potential issues.
        /// </summary>
        Warning
    }

    /// <summary>
    /// Represents a single validation error found during data validation.
    /// </summary>
    public class ValidationError
    {
        /// <summary>
        /// Gets the name of the table where the error occurred.
        /// </summary>
        public string TableName { get; }

        /// <summary>
        /// Gets the name of the field that failed validation.
        /// </summary>
        public string FieldName { get; }

        /// <summary>
        /// Gets the row key (primary key value) of the record with the error.
        /// </summary>
        public object? RowKey { get; }

        /// <summary>
        /// Gets a human-readable description of the validation error.
        /// </summary>
        public string Message { get; }

        /// <summary>
        /// Gets the severity level of this error.
        /// </summary>
        public ValidationSeverity Severity { get; }

        /// <summary>
        /// Gets the type of constraint that was violated.
        /// </summary>
        public string ConstraintType { get; }

        /// <summary>
        /// Gets the actual value that failed validation (if available).
        /// </summary>
        public object? ActualValue { get; }

        /// <summary>
        /// Creates a new validation error.
        /// </summary>
        public ValidationError(
            string tableName,
            string fieldName,
            object? rowKey,
            string message,
            string constraintType,
            ValidationSeverity severity = ValidationSeverity.Error,
            object? actualValue = null)
        {
            TableName = tableName;
            FieldName = fieldName;
            RowKey = rowKey;
            Message = message;
            ConstraintType = constraintType;
            Severity = severity;
            ActualValue = actualValue;
        }

        /// <inheritdoc/>
        public override string ToString()
        {
            var location = RowKey != null
                ? $"{TableName}[{RowKey}].{FieldName}"
                : $"{TableName}.{FieldName}";
            return $"[{Severity}] {location}: {Message}";
        }
    }

    /// <summary>
    /// Aggregates validation errors from data validation operations.
    /// </summary>
    public class ValidationResult
    {
        private readonly List<ValidationError> _errors = new();

        /// <summary>
        /// Gets all validation errors.
        /// </summary>
        public IReadOnlyList<ValidationError> Errors => _errors.AsReadOnly();

        /// <summary>
        /// Gets whether the validation passed (no errors).
        /// </summary>
        public bool IsValid => _errors.Count == 0;

        /// <summary>
        /// Gets the total number of errors.
        /// </summary>
        public int ErrorCount => _errors.Count;

        /// <summary>
        /// Gets errors filtered by severity.
        /// </summary>
        public IEnumerable<ValidationError> GetErrors(ValidationSeverity severity)
        {
            foreach (var error in _errors)
            {
                if (error.Severity == severity)
                    yield return error;
            }
        }

        /// <summary>
        /// Gets errors filtered by table name.
        /// </summary>
        public IEnumerable<ValidationError> GetErrorsForTable(string tableName)
        {
            foreach (var error in _errors)
            {
                if (error.TableName == tableName)
                    yield return error;
            }
        }

        /// <summary>
        /// Gets errors filtered by constraint type.
        /// </summary>
        public IEnumerable<ValidationError> GetErrorsByConstraint(string constraintType)
        {
            foreach (var error in _errors)
            {
                if (error.ConstraintType == constraintType)
                    yield return error;
            }
        }

        /// <summary>
        /// Adds a validation error to the result.
        /// </summary>
        public void AddError(ValidationError error)
        {
            _errors.Add(error);
        }

        /// <summary>
        /// Adds a validation error to the result.
        /// </summary>
        public void AddError(
            string tableName,
            string fieldName,
            object? rowKey,
            string message,
            string constraintType,
            ValidationSeverity severity = ValidationSeverity.Error,
            object? actualValue = null)
        {
            _errors.Add(new ValidationError(
                tableName, fieldName, rowKey, message, constraintType, severity, actualValue));
        }

        /// <summary>
        /// Merges another validation result into this one.
        /// </summary>
        public void Merge(ValidationResult other)
        {
            _errors.AddRange(other._errors);
        }

        /// <summary>
        /// Clears all errors.
        /// </summary>
        public void Clear()
        {
            _errors.Clear();
        }

        /// <inheritdoc/>
        public override string ToString()
        {
            if (IsValid)
                return "Validation passed: no errors.";

            return $"Validation failed with {_errors.Count} error(s):\n" +
                   string.Join("\n", _errors);
        }
    }

    /// <summary>
    /// Exception thrown when validation fails and the caller requested strict validation.
    /// </summary>
    public class ValidationException : Exception
    {
        /// <summary>
        /// Gets the validation result containing all errors.
        /// </summary>
        public ValidationResult Result { get; }

        /// <summary>
        /// Creates a new validation exception.
        /// </summary>
        public ValidationException(ValidationResult result)
            : base($"Data validation failed with {result.ErrorCount} error(s).")
        {
            Result = result;
        }

        /// <summary>
        /// Creates a new validation exception with a custom message.
        /// </summary>
        public ValidationException(ValidationResult result, string message)
            : base(message)
        {
            Result = result;
        }

        /// <inheritdoc/>
        public override string ToString()
        {
            return $"{base.ToString()}\n{Result}";
        }
    }

    /// <summary>
    /// Helper methods for validating field values against constraints.
    /// </summary>
    public static class ValidationHelpers
    {
        /// <summary>
        /// Validates that a string does not exceed the maximum length.
        /// </summary>
        public static bool ValidateMaxLength(string? value, int maxLength)
        {
            if (value == null) return true;
            return value.Length <= maxLength;
        }

        /// <summary>
        /// Validates that a value falls within the specified range (inclusive).
        /// </summary>
        public static bool ValidateRange<T>(T value, T min, T max) where T : IComparable<T>
        {
            return value.CompareTo(min) >= 0 && value.CompareTo(max) <= 0;
        }

        /// <summary>
        /// Validates that a nullable value falls within the specified range (inclusive).
        /// Returns true if value is null (nullable fields are valid when null).
        /// </summary>
        public static bool ValidateRangeNullable<T>(T? value, T min, T max) where T : struct, IComparable<T>
        {
            if (!value.HasValue) return true;
            return value.Value.CompareTo(min) >= 0 && value.Value.CompareTo(max) <= 0;
        }

        /// <summary>
        /// Validates that a string matches the specified regex pattern.
        /// </summary>
        public static bool ValidateRegex(string? value, string pattern)
        {
            if (value == null) return true;
            try
            {
                return Regex.IsMatch(value, pattern);
            }
            catch (ArgumentException)
            {
                // Invalid regex pattern - treat as validation failure
                return false;
            }
        }

        /// <summary>
        /// Validates that a string matches the specified regex pattern using a compiled Regex.
        /// </summary>
        public static bool ValidateRegex(string? value, Regex regex)
        {
            if (value == null) return true;
            return regex.IsMatch(value);
        }

        /// <summary>
        /// Validates that a non-optional field is not null.
        /// </summary>
        public static bool ValidateRequired<T>(T? value) where T : class
        {
            return value != null;
        }

        /// <summary>
        /// Validates that a non-optional value type field has been set.
        /// </summary>
        public static bool ValidateRequiredValue<T>(T? value) where T : struct
        {
            return value.HasValue;
        }

        /// <summary>
        /// Creates a MaxLength validation error.
        /// </summary>
        public static ValidationError MaxLengthError(
            string tableName,
            string fieldName,
            object? rowKey,
            int maxLength,
            int actualLength)
        {
            return new ValidationError(
                tableName,
                fieldName,
                rowKey,
                $"Value length ({actualLength}) exceeds maximum ({maxLength})",
                "MaxLength",
                ValidationSeverity.Error,
                actualLength);
        }

        /// <summary>
        /// Creates a Range validation error.
        /// </summary>
        public static ValidationError RangeError<T>(
            string tableName,
            string fieldName,
            object? rowKey,
            T min,
            T max,
            T actualValue)
        {
            return new ValidationError(
                tableName,
                fieldName,
                rowKey,
                $"Value ({actualValue}) is outside valid range [{min}, {max}]",
                "Range",
                ValidationSeverity.Error,
                actualValue);
        }

        /// <summary>
        /// Creates a Regex validation error.
        /// </summary>
        public static ValidationError RegexError(
            string tableName,
            string fieldName,
            object? rowKey,
            string pattern,
            string? actualValue)
        {
            return new ValidationError(
                tableName,
                fieldName,
                rowKey,
                $"Value does not match pattern: {pattern}",
                "Regex",
                ValidationSeverity.Error,
                actualValue);
        }

        /// <summary>
        /// Creates a ForeignKey validation error.
        /// </summary>
        public static ValidationError ForeignKeyError(
            string tableName,
            string fieldName,
            object? rowKey,
            string targetTable,
            object? foreignKeyValue)
        {
            return new ValidationError(
                tableName,
                fieldName,
                rowKey,
                $"Referenced record not found in {targetTable} (key: {foreignKeyValue})",
                "ForeignKey",
                ValidationSeverity.Error,
                foreignKeyValue);
        }

        /// <summary>
        /// Creates a Required field validation error.
        /// </summary>
        public static ValidationError RequiredError(
            string tableName,
            string fieldName,
            object? rowKey)
        {
            return new ValidationError(
                tableName,
                fieldName,
                rowKey,
                "Required field has null/empty value",
                "Required",
                ValidationSeverity.Error,
                null);
        }

        /// <summary>
        /// Creates a Unique constraint validation error.
        /// </summary>
        public static ValidationError UniqueError(
            string tableName,
            string fieldName,
            object? rowKey,
            object? duplicateValue)
        {
            return new ValidationError(
                tableName,
                fieldName,
                rowKey,
                $"Duplicate value found: {duplicateValue}",
                "Unique",
                ValidationSeverity.Error,
                duplicateValue);
        }
    }
}

File: static/typescript/validation.ts
// PolyGen Validation Support for TypeScript
// Provides validation utilities for checking data integrity after loading.

/**
 * Severity level for validation errors.
 */
export enum ValidationSeverity {
  /** Critical error that should prevent data usage. */
  Error = 'Error',
  /** Warning that may indicate potential issues. */
  Warning = 'Warning',
}

/**
 * Represents a single validation error found during data validation.
 */
export interface ValidationError {
  /** The name of the table where the error occurred. */
  tableName: string;
  /** The name of the field that failed validation. */
  fieldName: string;
  /** The row key (primary key value) of the record with the error. */
  rowKey: string;
  /** A human-readable description of the validation error. */
  message: string;
  /** The type of constraint that was violated. */
  constraintType: string;
  /** The severity level of this error. */
  severity: ValidationSeverity;
  /** The actual value that failed validation (if available). */
  actualValue?: unknown;
}

/**
 * Creates a validation error object.
 */
export function createValidationError(
  tableName: string,
  fieldName: string,
  rowKey: string,
  message: string,
  constraintType: string,
  severity: ValidationSeverity = ValidationSeverity.Error,
  actualValue?: unknown
): ValidationError {
  return {
    tableName,
    fieldName,
    rowKey,
    message,
    constraintType,
    severity,
    actualValue,
  };
}

/**
 * Formats a validation error as a string.
 */
export function formatValidationError(error: ValidationError): string {
  const location = error.rowKey
    ? `${error.tableName}[${error.rowKey}].${error.fieldName}`
    : `${error.tableName}.${error.fieldName}`;
  return `[${error.severity}] ${location}: ${error.message}`;
}

/**
 * Aggregates validation errors from data validation operations.
 */
export class ValidationResult {
  private _errors: ValidationError[] = [];

  /** Returns all validation errors. */
  get errors(): ReadonlyArray<ValidationError> {
    return this._errors;
  }

  /** Returns whether the validation passed (no errors). */
  get isValid(): boolean {
    return this._errors.length === 0;
  }

  /** Returns the total number of errors. */
  get errorCount(): number {
    return this._errors.length;
  }

  /** Adds a validation error to the result. */
  addError(error: ValidationError): void {
    this._errors.push(error);
  }

  /** Merges another validation result into this one. */
  merge(other: ValidationResult): void {
    this._errors.push(...other._errors);
  }

  /** Clears all errors. */
  clear(): void {
    this._errors = [];
  }

  /** Gets errors filtered by severity. */
  getErrorsBySeverity(severity: ValidationSeverity): ValidationError[] {
    return this._errors.filter((e) => e.severity === severity);
  }

  /** Gets errors filtered by table name. */
  getErrorsForTable(tableName: string): ValidationError[] {
    return this._errors.filter((e) => e.tableName === tableName);
  }

  /** Gets errors filtered by constraint type. */
  getErrorsByConstraint(constraintType: string): ValidationError[] {
    return this._errors.filter((e) => e.constraintType === constraintType);
  }

  /** Formats the validation result as a string. */
  toString(): string {
    if (this.isValid) {
      return 'Validation passed: no errors.';
    }
    const errorLines = this._errors.map(formatValidationError).join('\n  ');
    return `Validation failed with ${this._errors.length} error(s):\n  ${errorLines}`;
  }
}

/**
 * Error thrown when validation fails and the caller requested strict validation.
 */
export class ValidationException extends Error {
  public readonly result: ValidationResult;

  constructor(result: ValidationResult, message?: string) {
    super(message ?? `Data validation failed with ${result.errorCount} error(s).`);
    this.name = 'ValidationException';
    this.result = result;
  }
}

/**
 * Helper functions for validation.
 */
export const ValidationHelpers = {
  /**
   * Validates that a string does not exceed the maximum length.
   */
  validateMaxLength(value: string | null | undefined, maxLength: number): boolean {
    if (value == null) return true;
    return value.length <= maxLength;
  },

  /**
   * Validates that a value falls within the specified range (inclusive).
   */
  validateRange(value: number, min: number, max: number): boolean {
    return value >= min && value <= max;
  },

  /**
   * Validates that an optional value falls within the specified range.
   */
  validateRangeOptional(value: number | null | undefined, min: number, max: number): boolean {
    if (value == null) return true;
    return value >= min && value <= max;
  },

  /**
   * Validates that a string matches the specified regex pattern.
   */
  validateRegex(value: string | null | undefined, pattern: string): boolean {
    if (value == null) return true;
    try {
      const regex = new RegExp(pattern);
      return regex.test(value);
    } catch {
      return false;
    }
  },

  /**
   * Validates that a non-optional field is not null/undefined.
   */
  validateRequired<T>(value: T | null | undefined): value is T {
    return value != null;
  },

  /**
   * Creates a MaxLength validation error.
   */
  maxLengthError(
    tableName: string,
    fieldName: string,
    rowKey: string,
    maxLength: number,
    actualLength: number
  ): ValidationError {
    return createValidationError(
      tableName,
      fieldName,
      rowKey,
      `Value length (${actualLength}) exceeds maximum (${maxLength})`,
      'MaxLength',
      ValidationSeverity.Error,
      actualLength
    );
  },

  /**
   * Creates a Range validation error.
   */
  rangeError(
    tableName: string,
    fieldName: string,
    rowKey: string,
    min: number,
    max: number,
    actualValue: number
  ): ValidationError {
    return createValidationError(
      tableName,
      fieldName,
      rowKey,
      `Value (${actualValue}) is outside valid range [${min}, ${max}]`,
      'Range',
      ValidationSeverity.Error,
      actualValue
    );
  },

  /**
   * Creates a Regex validation error.
   */
  regexError(
    tableName: string,
    fieldName: string,
    rowKey: string,
    pattern: string,
    actualValue: string | null | undefined
  ): ValidationError {
    return createValidationError(
      tableName,
      fieldName,
      rowKey,
      `Value does not match pattern: ${pattern}`,
      'Regex',
      ValidationSeverity.Error,
      actualValue
    );
  },

  /**
   * Creates a ForeignKey validation error.
   */
  foreignKeyError(
    tableName: string,
    fieldName: string,
    rowKey: string,
    targetTable: string,
    foreignKeyValue: unknown
  ): ValidationError {
    return createValidationError(
      tableName,
      fieldName,
      rowKey,
      `Referenced record not found in ${targetTable} (key: ${foreignKeyValue})`,
      'ForeignKey',
      ValidationSeverity.Error,
      foreignKeyValue
    );
  },

  /**
   * Creates a Required field validation error.
   */
  requiredError(tableName: string, fieldName: string, rowKey: string): ValidationError {
    return createValidationError(
      tableName,
      fieldName,
      rowKey,
      'Required field has null/empty value',
      'Required',
      ValidationSeverity.Error,
      null
    );
  },

  /**
   * Creates a Unique constraint validation error.
   */
  uniqueError(
    tableName: string,
    fieldName: string,
    rowKey: string,
    duplicateValue: unknown
  ): ValidationError {
    return createValidationError(
      tableName,
      fieldName,
      rowKey,
      `Duplicate value found: ${duplicateValue}`,
      'Unique',
      ValidationSeverity.Error,
      duplicateValue
    );
  },
};
